### ISSUE TITLE
NIPS-2017-Attention is All you Need

### ISSUE THREAD

**Issue by BrambleXu** [BrambleXu/knowledge-graph-learning#102]
**一句话总结：**

Transformer的创世文。比recurrent和conventional NN更快(能并行)，能更好学习长距离关系。

**资源：**

- [pdf](<https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf>)
- code: [Tensor2Tensor](<https://github.com/tensorflow/tensor2tensor>), [Fairseq](<https://github.com/pytorch/fairseq>)
- [paper-with-code](<https://paperswithcode.com/paper/attention-is-all-you-need>)
- [The Annotated Transformer](<http://nlp.seas.harvard.edu/2018/04/03/attention.html>), [中文翻译](<https://www.jiqizhixin.com/articles/Synced-github-implement-project-machine-translation-by-transformer>)

**关键字：**

- dataset: 
- Transformer, based solely on attention mechanisms
- Scaled Dot-Product Attention
- Multi-head attention

**笔记：**

# 1 Introduction

encoder-decoder architectures:

- LSTM
- GRU

这种模型的缺点：无法并行处理。对于长序列来说，非常耗时。

而Attention mechanisms经常作为上面模型的一部分被使用。

这篇文章的工作是提出了Transformer，避免了recurrence，完全依赖于注意力机制来draw global dependencies between input and output. 

# 2 Background

为了reducing sequential computation，很多模型(Extended Neural GPU
[20], ByteNet [15] and ConvS2S [8])都采用CNN作为计算单元，因为CNN可以并行，速度快。但是这些基于CNN的模型，对于序列中长距离关系的学习效果不好。difficult to learn dependencies between distant positions (Gradient flow in recurrent nets: the difficulty of learning long-term dependencies, 2001).  但是在Transformer中，序列计算的次数可以被缩减到常数级别，通过averaging attention-weighted positions，但副作用是牺牲了一部分效果，这里引入了Multi-Head Attention来抵消这部分副作用。

Self-attention, sometimes called intra-attention is an attention mechanism relating different positions of a single sequence in order to compute a representation of the sequence. 这个Self-attention的机制是计算一个序列中不同的位置，来得到sequence representation. 其有效性在其他task中也被证明了。Self-attention has been used successfully in a variety of tasks including reading comprehension,  bstractive summarization, textual entailment and learning task-independent sentence representations [4, 22, 23, 19].

End-to-end memory networks则是基于recurrent attention mechanism instead of sequencealigned recurrence and have been shown to perform well on simple-language question answering and
language modeling tasks.(End-to-end memory networks 2015)

上面的模型都是把attention作为一部分来用。Transformer则是完全基于self-attention来计算representations of its input and output without using sequencealigned RNNs or convolution。灵感来源于下面的文章，有提到它的优点：

- Neural GPUs learn algorithms，(ICLR), 2016.
- Neural machine translation in linear time. arXiv  2017
- Convolutional sequence to sequence learning, arXiv  2017

# 3 Model Architecture

大部分序列转导模型( sequence transduction models)都是encoder-decoder structure。encoder把一个token序列变为 continuous representations $z$. 有了z之后，the decoder then generates an output sequence (y1, ..., ym) of symbols one element at a time (decoder根据生成一个output sequence，每次只产生一个)。每个step，模型是auto-regressive (自我回归)的，即把前一个step产生的symbol作为额外的input来产生下一个symbol。

Transfromer也遵循了这样的结构，使用了stacked self-attention and point-wise, fully connected layers for both the encoder and decoder, 结构图如下：

![image-20190411110011258](https://ws1.sinaimg.cn/large/006tNc79gy1g1ygpja7ruj30ee0l3mzk.jpg)

## 3.1 Encoder and Decoder Stacks

- a residual connection ？ (Deep residual learning for image recognition 2016 CVPR)

## 3.2 Attention

An attention function can be described as mapping a query and a set of key-value pairs to an output, where the query, keys, values, and output are all vectors. The output is computed as a weighted sum of the values, where the weight assigned to each value is computed by a compatibility function of the query with the corresponding key.

这里的query和key-value pair该如何理解？

### 3.2.1 Scaled Dot-Product Attention

We call our particular attention "Scaled Dot-Product Attention"

![image-20190411110536561](https://ws4.sinaimg.cn/large/006tNc79gy1g1ygv4evspj30qk0fcdi0.jpg)

The input consists of queries and keys of dimension $d_k$, and values of dimension $d_v$. 

![image-20190411110930619](https://ws4.sinaimg.cn/large/006tNc79gy1g1ygz6g2tnj30qr0akacs.jpg)

### 3.2.2 Multi-Head Attention

Instead of performing a single attention function with dmodel-dimensional keys, values and queries, we found it beneficial to linearly project the queries, keys and values h times with different, learned linear projections to dk, dk and dv dimensions, respectively。 

### 3.2.3 Applications of Attention in our Model

![image-20190411111335421](https://ws2.sinaimg.cn/large/006tNc79gy1g1yh3f728zj30qf0ca421.jpg)

## 3.3 Position-wise Feed-Forward Networks

？

## 3.4 Embeddings and Softmax

We use learned embeddings to convert the input tokens and output tokens to vectors of dimension $d_{mode}$。

## 3.5 Positional Encoding

因为Transformer没有recurrence和convolution，为了利用序列里的位置顺序关系，比如加入这些信息。we must inject some information about the relative or absolute position of the tokens in the sequence. we add "positional encodings" to the input embeddings at the bottoms of the encoder and decoder stacks. 

# 4 Why Self-Attention

为什么Self-Attention比recurrent和convolutional layer更好。

- One is the total computational complexity per layer. 

- Another is the amount of computation that canbe parallelized, as measured by the minimum number of sequential operations equired

- The third is the path length between long-range dependencies in the network。 ？The shorter these paths between any combination of positions in the input and output sequences, the easier it is to learn long-range dependencies (没看懂，相关论文是Gradient flow in
  recurrent nets: the difficulty of learning long-term dependencies, 2001)  (**其实RE的cross-sentence问题完全可以看成是long range dependency问题来解决，所以直觉上讲，如果transformer在long range dependency上表现效果好的话，那么在cross-sentence RE上理应也有不错的效果)

  

# 6 Results

## 6.1 Machine Translation

![image-20190411113344654](https://ws2.sinaimg.cn/large/006tNc79gy1g1yhoe3ig5j30q90bjjuk.jpg)

## 6.2 Model Variations

![image-20190411113406427](https://ws2.sinaimg.cn/large/006tNc79gy1g1yhorv7y8j30ps0k9ae0.jpg)
